### PR TITLE
Line reader

### PR DESCRIPTION
**Pull request by yingsu00**
Throw IOException when the text line is too big and
1) The capacity of the Text object is over the VM limit for the array size;
2) The bytes to store into the Text object is over the maxLineLength limit;
3) The consumed bytes is over the maxBytesToConsume.

Case 1) would cause OOM, and the fix (the same as the second commit in this PR)
was submitted as the following PR to Apache Hadoop:
https://github.com/apache/hadoop/pull/414

The fix for case 2) and 3) was behavior change so that when the line is over the configured
maxLineLength limit (suppose it's less than the VM array size limit), the LineReader 
throws IOException so that the query would fail loudly. The original LineReader behavior 
was to silently discard the content for a line after the maxLineLength limit.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -366,24 +367,6 @@
...
-                        <phase>verify</phase>
-                        <goals>
-                            <goal>check</goal>
-                        </goals>
-                    </execution>
-                </executions>
-                <configuration>
-                    <failBuildInCaseOfConflict>true</failBuildInCaseOfConflict>
-                </configuration>
-            </plugin>
```
Comment (kokosing) [inline pom.xml]: why is this removed?

**Comment by dain** [comment]
Can you submit a PR (or issue) to Hadoop to get this change merged upstream, and then reference it in the code... basically, say, "Remove this class once xxx is resolved in Hadoop"

File: pom.xml (hunk 1 above)
Comment (yingsu00) [reply pom.xml]: @kokosing The LineReader class is duplicated and this plugin would throw error when building.

Comment (kokosing) [reply pom.xml]: But you excluded this file here: https://github.com/prestodb/presto-hadoop-apache2/pull/30/files#diff-600376dffeb79835ede4a0b285078036R546.

This plugin was added here for a reason to catch that situation. If class is duplicated, then which one will be used?

Comment (yingsu00) [reply pom.xml]: That exclude doesn't work. I don't see what this maven-duplicate-finder-plugin is used for. There are no other duplicated classes. As a similar example, the presto-hive-apache pom.xml doesn't include this maven-duplicate-finder-plugin as well. However I added it back but changed the configuration to
<configuration>
<failBuildInCaseOfEqualContentConflict>true</failBuildInCaseOfEqualContentConflict>
</configuration>
@electrum what do you think?

**Comment by yingsu00** [comment]
@kokosing Regarding the maven-duplicate-finder-plugin, as @electrum explained, is a similar plugin that does the same thing of maven-shade-plugin, and we only need one. I have again removed maven-duplicate-finder-plugin and kept the maven-shade-plugin with the exclude of LineReader class.

File: pom.xml (hunk 1 above)
Comment (electrum) [reply pom.xml]: The shade plugin does a duplicate check itself. Unfortunately, it's only a warning and doesn't fail the build, but shading is complicated and doesn't change often, so it seems reasonable that anyone working on it or reviewing can review the build output. (there are lots of other things you have to review manually when shading, this is one of them)

Having the duplicate checker is problematic because it doesn't use the exclusion config from the shade plugin. It has to be configured separately, and can be misleading since you might only configure the duplicate checker and not the shade plugin and would be lead to believe the build was safe.

Hence, my recommendation is to remove it.

File: pom.xml [hunk 2]
```diff
@@ -340,6 +340,7 @@
             <version>6.2.1</version>
             <scope>test</scope>
         </dependency>
+
```
Comment (electrum) [inline pom.xml]: Nit: this looks like an accidental change

File: pom.xml [hunk 3]
```diff
@@ -366,24 +367,6 @@
                 </configuration>
             </plugin>
 
-            <plugin>
```
Comment (electrum) [inline pom.xml]: Let's do this removal first, as a separate commit. Something like
```
Remove duplicate finder plugin

The shade plugin already checks for duplicate classes.

File: src/main/java/org/apache/hadoop/util/LineReader.java [hunk 4]
```diff
@@ -244,6 +244,10 @@ private int readDefaultLine(Text str, int maxLineLength, int maxBytesToConsume)
             int appendLength = readLength - newlineLength;
             if (appendLength > maxLineLength - txtLength) {
                 appendLength = maxLineLength - txtLength;
+                if (appendLength > 0) {
+                    // We want to fail the query when the line length is over the limit.
```
Comment (electrum) [inline src/main/java/org/apache/hadoop/util/LineReader.java]: This comment is out of place here, since this is a generic library. We could say *"fail the read"* instead. The same goes for the other comments.

**Comment by electrum** [comment]
Can you add test for `LineReader` to cover the new behavior? It should be easy to create one using `ByteArrayInputStream` as the input source. Ideally, the test should cover all of the places where the exception is thrown (run with coverage in your IDE and make sure all of the throw statements are hit).

File: src/main/java/org/apache/hadoop/util/LineReader.java [hunk 5]
```diff
@@ -41,6 +41,10 @@
 @InterfaceAudience.LimitedPrivate({"MapReduce"})
 @InterfaceStability.Unstable
 public class LineReader implements Closeable {
+    // Limitation for array size is VM specific. Current HotSpot VM limitation
+    // for array size is Integer.MAX_VALUE - 5 (2^31 - 1 - 5).
+    // Integer.MAX_VALUE - 8 should be safe enough.
```
Comment (electrum) [inline src/main/java/org/apache/hadoop/util/LineReader.java]: The comment says `- 8` but the code says `- 9`.

File: src/main/java/org/apache/hadoop/util/LineReader.java [hunk 6]
```diff
@@ -240,12 +246,19 @@ private int readDefaultLine(Text str, int maxLineLength, int maxBytesToConsume)
                 appendLength = maxLineLength - txtLength;
             }
             if (appendLength > 0) {
+                int newTxtLength = txtLength + appendLength;
+                if (str.getBytes().length < newTxtLength && Math.max(newTxtLength, txtLength << 1) > MAX_ARRAY_SIZE) {
```
Comment (electrum) [inline src/main/java/org/apache/hadoop/util/LineReader.java]: Do we need to worry about overflow of `txtLength << 1`? Same question for the one below

File: src/main/java/org/apache/hadoop/util/LineReader.java [hunk 7]
```diff
@@ -240,12 +246,19 @@ private int readDefaultLine(Text str, int maxLineLength, int maxBytesToConsume)
...
+                    throw new IOException("Too many bytes before newline:  " + newTxtLength);
+                }
                 str.append(buffer, startPosn, appendLength);
-                txtLength += appendLength;
+                txtLength = newTxtLength;
             }
         } while (newlineLength == 0 && bytesConsumed < maxBytesToConsume);
 
         if (bytesConsumed > Integer.MAX_VALUE) {
+
```
Comment (electrum) [inline src/main/java/org/apache/hadoop/util/LineReader.java]: Accidental change

File: pom.xml (hunk 1 above)
Comment (kokosing) [reply pom.xml]: Thanks, for explanation.

File: src/main/java/org/apache/hadoop/util/LineReader.java (hunk 6 above)
Comment (yingsu00) [reply src/main/java/org/apache/hadoop/util/LineReader.java]: This is to simulate the behavior of Text.setCapacity() method. It would try to resize Text.bytes to Math.max(newTxtLength, txtLength << 1) when keepData is true, which is the case for append().

Comment (yingsu00) [reply src/main/java/org/apache/hadoop/util/LineReader.java]: In such case Math.max(newTxtLength, txtLength << 1) will be newTxtLength which is <= maxLineLength and won't overflow. This is the same logic in Text.setCapacity().

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    @Test
+    public void testDefaultReaderMaxBytesConsumed()
+    {
+        byte[] input = "Hello world! Goodbye world!\n".getBytes(UTF_8);
+        InputStream in = new ByteArrayInputStream(input);
+        // Set the LineReader internal read buffer size 4 bytes
+        LineReader reader = new LineReader(in, 4);
+        Text str = new Text();
+        try {
+            reader.readLine(str, 0, 10);
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: The test will succeed if it doesn't throw an exception. We need to add this:
```java
fail("expected exception");
```
Same goes for the other tests.

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        catch (IOException e) {
+            // It should be 3 reads of 4 bytes each, so the final bytesConsumed is 12
+            assertEquals(e.getMessage(), "Too many bytes before newline: " + 12);
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: Let's inline the string concat to make it easier to read:
```java
assertEquals(e.getMessage(), "Too many bytes before newline: 12");
```
Same goes for the other tests.

**Comment by electrum** [comment]
Can you also add a test for `LineReader` that passes?

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+
+public class TestLineReader {
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: Nit: formatting for this is off

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        byte[] input = "Hello world! Goodbye world!\n".getBytes(UTF_8);
+        InputStream in = new ByteArrayInputStream(input);
+        // Set the LineReader internal read buffer size 4 bytes
+        LineReader reader = new LineReader(in, 4);
+        Text str = new Text();
+        try {
+            // There should not be any bytes read into str because the maxLineLength is 0
+            reader.readLine(str, 0, 30);
+            assertEquals(str.getLength(), 0);
+            assertEquals(str.getBytes(), "".getBytes());
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: This second test will never be hit, because the first one would fail first if the string is non-empty. It would be a bit better to replace both tests with this:
```java
assertEquals(str, new Text());
```
This will use the `equals` method of `Text`, which works correctly. The advantage is that on a failure, it will show the results of `toString` for the actual and expected values, which makes it easier to debug (compared to just showing the length is different).

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        }
+        catch (IOException e) {
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: This catch is not needed. Just declare the checked exception for the test method. If it throws, the test will fail.

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        InputStream in = new ByteArrayInputStream(input);
+        // Set the LineReader internal read buffer size 4 bytes
+        LineReader reader = new LineReader(in, 4);
+        Text str = new Text();
+        try {
+            reader.readLine(str, 30, 30);
+            // The LineReader does not store the new line character into str,
+            // and the str was doubled 3 times so its size is 32, with 5 0-valued
+            // bytes at the end, so we need to compare just the first 27 characters.
+            assertEquals(str.getLength(), 27);
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: I would write this as
```java
String input = "Hello world! Goodbye world!\n";
...
assertEquals(str, new Text(input));
```
You want to make tests as possible, both so they are easy to read and to avoid bugs in tests.

File: src/test/java/com/facebook/presto/hadoop/TestLineReader.java [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    {
+        byte[] input = "Hello world! Goodbye world!\n".getBytes(UTF_8);
+        InputStream in = new ByteArrayInputStream(input);
+        // Set the LineReader internal read buffer size 4 bytes
+        LineReader reader = new LineReader(in, 4);
+        Text str = new Text();
+        try {
+            // The read should fail because the line length is greater than the maxBytesToConsume
+            reader.readLine(str, 0, 10);
+            fail("Expected exception " + IOException.class.getSimpleName());
```
Comment (electrum) [inline src/test/java/com/facebook/presto/hadoop/TestLineReader.java]: Sorry, I actually meant to just write
```java
fail("expected exception");
```
Alternatively, we could write
```java
fail("expected IOException");
```
While you often want to avoid hard-coding class names, in this case, there's no value for doing it given that "IOException" is part of the JDK and will never be renamed. Doing it that way here makes the code more complex and thus harder to read.
